Add new info.Info interface and add minimal package yaml parser

mvo5 opened the pull request:
This branch is motivated by the need to have the "type" of a snap available in ubuntu-device-flash so that I can infer if its a kernel or an OS snap. When those are detected u-d-f needs to setup the boot variables accordingly.

It also moves pkg.File:Open into its own package to avoid circular dependencies. If you prefer that in a separate branch I can do that.

I think this change has some potential and we could use it as a start to get packageYaml out of the main "snappy" package.

mvo5 commented:
retest this please

chipaca commented:
I like what you're doing here. I don't like the name `opener`, because once you've imported it's not clear that all it opens are packages :-) but I don't have a better suggestion.

mvo5 commented:
Thanks @chipaca ! I renamed it to snapfile.Open now and moved the pkg.File in there too (but I can easily revert that last commit if you prefer it as pkg.File).

mvo5 commented:
Fwiw, the integration tests shold be ok again once bugfix/oem-compat is merged into master.

mvo5 commented:
Once this has landed we can land a branch for ubuntu-device-flash in 16.04 that can create all-snap images (and only all-snap images).

mvo5 commented:
And once we have u-d-f in 16.04 with all-snap support the integration-tests can be updated to all-snap images too.

chipaca commented on pkg/info/info_package_yaml.go:
```diff
@@ -0,0 +1,66 @@
+	Type    pkg.Type
+}
+
+// SnapInfoPackageYaml implements the meta/snap.yaml data
```
`snap.yaml` already?

mvo5 replied:
Not yet, but soon :) the whole layout is anticipating snap.yaml

chipaca commented:
:+1:

niemeyer commented:
For the record, we discussed this online and @mvo5 is working on some changes right now.

mvo5 commented:
The branch is updated now based on the feedback from @niemeyer (thanks!). Its a bit long unfortunately mostly due to the renames. Feedback (very) welcome.

niemeyer commented on snap/file.go:
```diff
@@ -43,6 +38,22 @@ type File interface {
+}
+
+// Backend implements a specific snap format
+type Backend struct {
```
Looks like this should be private for now. I'd also take the chance we have a good term for it after "file format" (e.g. "snapFormat" and "RegisterFormat", or something similar of your liking) and reserve the generic "backend" wildcard term for something else.

niemeyer commented on snap/info_package_yaml.go:
```diff
@@ -0,0 +1,38 @@
+)
+
+// NewFromPackageYaml creates a new info based on the given packageYaml
+func NewFromPackageYaml(yamlData []byte) (*Info, error) {
```
This should probably be `snap.InfoFromPackageYaml` now that it sits under the `snap` package (it's not a "new snap").

niemeyer commented:
Just a couple of trivials now.. really like the outcome, though! LGTM

niemeyer commented:
This changed enough that would be good for @chipaca to have another look.

chipaca commented on snap/clickdeb/deb.go:
```diff
@@ -48,6 +49,12 @@ var (
 
+func init() {
+	snap.RegisterBackend([]byte("!<arch>\ndebian"), func(path string) (snap.File, error) {
+		return Open(path)
```
why isn't this
```go
	snap.RegisterBackend([]byte("!<arch>\ndebian"), Open)
```
?

chipaca commented on snappy/click.go:
```diff
@@ -1,7 +1,7 @@
 
 /*
- * Copyright (C) 2014-2015 Canonical Ltd
+ * Copyyright (C) 2014-2015 Canonical Ltd
```
ccooppyyrriigghhtt
also maybe +2016.

chipaca commented:
:+1:
